### PR TITLE
[Enhancement] support export sink in pipeline engine

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

major changes:
1.  support export sink in pipeline engine;
2. some sinks with sync io have a similar working mode, so I add the abstraction SinkIOBuffer abstraction to make it easier to use. 
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 8 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|7|00.00%|[287, 288, 289, 290, 291, 292, 316]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 8 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|7|00.00%|[287, 288, 289, 290, 291, 292, 316]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 8 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|7|00.00%|[287, 288, 289, 290, 291, 292, 316]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|6|00.00%|[287, 288, 289, 290, 291, 292]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|6|00.00%|[287, 288, 289, 290, 291, 292]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

File: be/src/exec/pipeline/sink/export_sink_operator.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        close(_state);
+        return;
+    }
+
+    if (_file_builder == nullptr) {
+        int query_timeout = _state->query_options().query_timeout;
+        int timeout_ms = query_timeout > 3600 ? 3600000 : query_timeout * 1000;
+        if (Status status = _open_file_writer(timeout_ms); !status.ok()) {
+            set_io_status(status);
+            close(_state);
```
Comment (fzhedu) [inline be/src/exec/pipeline/sink/export_sink_operator.cpp:84]: what if query_timeout = 0?

File: be/src/exec/pipeline/sink/export_sink_operator.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        if (_t_export_sink.__isset.use_broker && !_t_export_sink.use_broker) {
+            ASSIGN_OR_RETURN(auto fs, FileSystem::CreateUniqueFromString(file_path, FSOptions(&_t_export_sink)));
+            ASSIGN_OR_RETURN(output_file, fs->new_writable_file(options, file_path));
+            break;
+        } else {
+            const TNetworkAddress& broker_addr = _t_export_sink.broker_addresses[0];
+            BrokerFileSystem fs_broker(broker_addr, _t_export_sink.properties, timeout_ms);
+            ASSIGN_OR_RETURN(output_file, fs_broker.new_writable_file(options, file_path));
+            break;
+        }
```
Comment (fzhedu) [inline be/src/exec/pipeline/sink/export_sink_operator.cpp:124]: move the two break after L:124

File: be/src/exec/pipeline/sink/export_sink_operator.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    const std::vector<ExprContext*> _output_expr_ctxs;
+    std::unique_ptr<FileBuilder> _file_builder;
+};
+
+Status ExportSinkIOBuffer::prepare(RuntimeState* state, RuntimeProfile* parent_profile) {
+    bool expected = false;
+    if (!_is_prepared.compare_exchange_strong(expected, true)) {
+        return Status::OK();
+    }
+    _state = state;
```
Comment (fzhedu) [inline be/src/exec/pipeline/sink/export_sink_operator.cpp:46]: would the state be null?

Comment (silverbullet233) [reply be/src/exec/pipeline/sink/export_sink_operator.cpp:46]: > would the state be null?

no.

File: be/src/exec/pipeline/sink/export_sink_operator.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/sink/export_sink_operator.cpp:84]: > what if query_timeout = 0?

If the user actively sets it to 0, the query will fail immediately

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|6|00.00%|[287, 288, 289, 290, 291, 292]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|6|00.00%|[287, 288, 289, 290, 291, 292]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/exec/pipeline/sink/export_sink_operator.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+namespace starrocks {
+
+class RowDescriptor;
+
+namespace pipeline {
+
+class ExportSinkIOBuffer;
+class FragmentContext;
+
+class ExportSinkOperator : public Operator {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/sink/export_sink_operator.h:20]: final

Comment (silverbullet233) [inline be/src/exec/pipeline/sink/export_sink_operator.h:20]:
```suggestion
class ExportSinkOperator final : public Operator {
```

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by fzhedu — APPROVED** [review]
LGTM

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|6|00.00%|[287, 288, 289, 290, 291, 292]|
|:large_blue_circle:|com/starrocks/planner/ExportSink.java|0|1|00.00%|[123]|

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**